pkg/cmd/cli/cmd/debug.go

```diff
@@ -0,0 +1,565 @@
+  $ %[1]s dc/test -c second -- /bin/env
+
+  # See the pod that would be created to debug
+  $ %[1]s dc/test -o yaml`
+
+	debugPodLabelName = "debug.openshift.io/name"
+
+	debugPodAnnotationSourceContainer = "debug.openshift.io/source-container"
+	debugPodAnnotationSourceResource  = "debug.openshift.io/source-resource"
+)
+
+// NewCmdDebug creates a command for debugging pods.
+func NewCmdDebug(fullName string, f *clientcmd.Factory, in io.Reader, out, errout io.Writer) *cobra.Command {
+	options := &DebugOptions{
+		Timeout: 30 * time.Second,
+		Command: []string{"/bin/sh"},
+		Attach: kcmd.AttachOptions{
+			In:    in,
+			Out:   out,
+			Err:   errout,
+			TTY:   true,
+			Stdin: true,
+
+			Attach: &kcmd.DefaultRemoteAttach{},
+		},
+		LogsForObject: f.LogsForObject,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "debug RESOURCE/NAME [ENV1=VAL1 ...] [-c CONTAINER] [-- COMMAND]",
+		Short:   "Launch a new instance of a pod for debugging",
+		Long:    debugLong,
+		Example: fmt.Sprintf(debugExample, fmt.Sprintf("%s debug", fullName)),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Complete(cmd, f, args, in, out, errout))
+			kcmdutil.CheckErr(options.Validate())
+			kcmdutil.CheckErr(options.Debug())
+		},
+	}
+
+	// TODO: when T is deprecated use the printer, but keep these hidden
+	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml|wide|name|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=... See golang template [http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template [http://releases.k8s.io/HEAD/docs/user-guide/jsonpath.md].")
+	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
+	cmd.Flags().String("template", "", "Template string or path to template file to use when -o=go-template, -o=go-template-file. The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].")
+	cmd.Flags().Bool("no-headers", false, "When using the default output, don't print headers.")
+	cmd.Flags().MarkHidden("no-headers")
+	cmd.Flags().String("sort-by", "", "If non-empty, sort list types using this field specification.  The field specification is expressed as a JSONPath expression (e.g. 'ObjectMeta.Name'). The field in the API resource specified by this JSONPath expression must be an integer or a string.")
+	cmd.Flags().MarkHidden("sort-by")
+	cmd.Flags().Bool("show-all", true, "When printing, show all resources (default hide terminated pods.)")
+	cmd.Flags().MarkHidden("show-all")
+
+	cmd.Flags().BoolVarP(&options.NoStdin, "no-stdin", "I", options.NoStdin, "Force a pseudo-terminal to be allocated")
+	cmd.Flags().BoolVarP(&options.ForceTTY, "tty", "t", false, "Force a pseudo-terminal to be allocated")
+	cmd.Flags().BoolVarP(&options.DisableTTY, "no-tty", "T", false, "Disable pseudo-terminal allocation")
```

Comment:
Incidentally, -t and -T exactly match ssh

On Fri, Feb 19, 2016 at 3:55 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> the default behavior is like RSH - use the console settings automatically
> (it may or may not be a tty).  So we have -t to force use and -T to force
> not use.
>
> On Fri, Feb 19, 2016 at 3:54 PM, David Eads <notifications@github.com>
> wrote:
>
>> In pkg/cmd/cli/cmd/debug.go
>> <https://github.com/openshift/origin/pull/7294#discussion_r53516859>:
>>
>> > +	}
>> > +
>> > +	// TODO: when T is deprecated use the printer, but keep these hidden
>> > +	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml|wide|name|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=... See golang template [http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template [http://releases.k8s.io/HEAD/docs/user-guide/jsonpath.md].")
>> > +	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
>> > +	cmd.Flags().String("template", "", "Template string or path to template file to use when -o=go-template, -o=go-template-file. The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].")
>> > +	cmd.Flags().Bool("no-headers", false, "When using the default output, don't print headers.")
>> > +	cmd.Flags().MarkHidden("no-headers")
>> > +	cmd.Flags().String("sort-by", "", "If non-empty, sort list types using this field specification.  The field specification is expressed as a JSONPath expression (e.g. 'ObjectMeta.Name'). The field in the API resource specified by this JSONPath expression must be an integer or a string.")
>> > +	cmd.Flags().MarkHidden("sort-by")
>> > +	cmd.Flags().Bool("show-all", true, "When printing, show all resources (default hide terminated pods.)")
>> > +	cmd.Flags().MarkHidden("show-all")
>> > +
>> > +	cmd.Flags().BoolVarP(&options.NoStdin, "no-stdin", "I", options.NoStdin, "Force a pseudo-terminal to be allocated")
>> > +	cmd.Flags().BoolVarP(&options.ForceTTY, "tty", "t", false, "Force a pseudo-terminal to be allocated")
>> > +	cmd.Flags().BoolVarP(&options.DisableTTY, "no-tty", "T", false, "Disable pseudo-terminal allocation")
>>
>> I don't understand why I have both these flags tty and no-tty
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/7294/files#r53516859>.
>>
>
>